File: Rydberg ion engine/Matlab/quench_delta_ryd/quench.m
function a = quench(t, T, a_max, a_min)
%UNTITLED2 Summary of this function goes here
%   Detailed explanation goes here

%angular coefficient
ang = 4.0*(a_max - a_min)/T;

%linear coefficients
lin1 = a_max - 2.0*a_min;
lin2 = 4.0*a_max - 3.0*a_min;


%in order to understand the mechanism behind this routine, remember that
%mod(t, T) = t - nT where n is a integer

%defining the domain
%the first one 0 < t < pi/2 
d1 = (0 <= mod(t, T) & mod(t, T) < T/4);
a(d1) = a_min;

%the second one pi/2 < t < pi
d2 = (T/4 <= mod(t, T) & mod(t, T) < T/2);
a(d2) = ang*mod(t(d2), T) - lin1;

%the third one pi < t < 3*pi/2
d3 = (T/2 <= mod(t, T) & mod(t, T) < 3*T/4);
a(d3) = a_max;

%the fourth one 0 < t < pi/2
d4 = (3.0*T/4.0 <= mod(t, T) & mod(t, T) <= T);
a(d4) = lin2 - ang*mod(t(d4), T);

end

File: Rydberg ion engine/Python/time_dependent/quench.py
from qutip import*
import numpy as np
import matplotlib.pyplot as plt

def quench(t, T, a_max, a_min):

    #angular coefficient 
    ang = 4.0*(a_max - a_min)/T

    #linear coefficients
    lin1 = a_max - 2.0*a_min
    lin2 = 4.0*a_max - 3.0*a_min

    a1 = lambda t: a_min
    a2 = lambda t: ang*(t % T)  - lin1
    a3 = lambda t: a_max
    a4 = lambda t: lin2 - ang*(t % T)

    return np.piecewise(t, [(0.0 <= t % T)*(t % T < T/4.0),
    (T/4.0 <= t % T)*(t % T < T/2.0),
    (T/2.0 <= t % T)*(t % T < 3.0*T/4.0),
    (3.0*T/4.0 <= t % T)*(t % T < T)],
    [a1, a2, a3, a4]) 

#tspan = np.linspace(0, 100, 10000)
#plt.plot(tspan, quench(tspan, 4*np.pi, 1, -1))
#plt.show()
